### ISSUE TITLE
Add a message that prints during fitting if terms are aliased

### ISSUE THREAD

**Issue by gavinsimpson** [vegandevs/vegan#682]
Many users wonder why they have fewer constrained axes than constraints they include in their model. We currently include a note in the output when printing the model, but seemingly users either miss this or do not understand it.

Instead, print a message about the aliasing when fitting the model/ordination, in addition to the existing note in the printed output.

See discussion in #671, in which @jarioksa suggests:

* [x] `ordConstrain` should issue a `message()` of aliased variables (and preferably in plain English: the current one of aliasing sounds complicated).
* [x] `ordConstrained` should have a `message()` instead of current `warning()` of overfitted models and missing residual component.
* [x] `print.cca` should print a text of missing residual component (and overfitted) model in addition to the current text of aliased redundant constraints and/or conditions.

**Comment by gavinsimpson** [comment]
@jarioksa One infelicity of printing a `message()` (or just printing something to the console) is that functions that fit a model repeatedly generate messages:

```r
> drop1(mcca,  test = "permutation", permutations = per)

Some constraints or conditions were aliased because they were redundant. This can
happen if terms are linearly dependent (collinear): ‘Manure^4’

Some constraints or conditions were aliased because they were redundant. This can
happen if terms are linearly dependent (collinear): ‘Manure^4’
                      Df  AIC    F Pr(>F)  
<none>                   87.9              
Condition(Management)  2 87.9              
Manure                 3 86.5 1.03   0.86  
A1                     1 89.0 2.00   0.04 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
```

We'll need to wrap some calls in some functions like `drop1.cca` in `suppressMessages()`.

Right now I know the following which need attention:

* [ ] `add11.cca()`
* [ ] `drop1.cca()`
* [ ] `anovaCCAbymargin()`
* [ ] `anovaCCAbyaxis()`

**Comment by gavinsimpson** [comment]
I think using `suppressMessages()` in `add1` and `drop1` methods is OK because adding terms should never get into a situation where you add a linearly dependent term and dropping terms won't create a problem unless the model the user fitted already had aliased terms and thus will already have seen the message.

**Comment by gavinsimpson** [comment]
@jarioksa

OK, I think I have something ready for your input:

### message emitted when fitting model

```r
> mod <- cca(dune ~ . + Condition(Manure), dune.env)

Some constraints or conditions were aliased because they were redundant. This can
happen if terms are linearly dependent (collinear): ‘ManagementNM’, ‘Manure.L’,
‘Manure.Q’, ‘Manure.C’, ‘Manure^4’
```

### output when printing the model

```r
> mod
Call: cca(formula = dune ~ A1 + Moisture + Management + Use + Manure +
Condition(Manure), data = dune.env)

-- Model Summary --

              Inertia Proportion Rank
Total          2.1153     1.0000     
Conditional    0.6116     0.2891    4
Constrained    0.8916     0.4215    8
Unconstrained  0.6121     0.2894    7

Inertia is scaled Chi-square

-- Note --

Some constraints or conditions were aliased because they were redundant. This
can happen if terms are linearly dependent (collinear): ‘ManagementNM’,
‘Manure.L’, ‘Manure.Q’, ‘Manure.C’, ‘Manure^4’

-- Eigenvalues --

Eigenvalues for constrained axes:
  CCA1   CCA2   CCA3   CCA4   CCA5   CCA6   CCA7   CCA8 
0.3318 0.1585 0.1095 0.1004 0.0832 0.0557 0.0286 0.0240 

Eigenvalues for unconstrained axes:
    CA1     CA2     CA3     CA4     CA5     CA6     CA7 
0.27237 0.10876 0.08975 0.06305 0.03489 0.02529 0.01798 

```

Note that I have made more significant changes to `print.cca` to facilitate printing a more informative notification. When I added the aliasing note the important results disappeared, especially if there were more notes to print (e.g. deleted samples/species). So I decided to add a heading for the sections to highlight the results. I'm not wedded to `-- Heading --` notation - suggestions for alternatives welcome.

I have fixed the methods mentioned above that would refit a model and thus cause more messages to be printed. So we just need to agree / OK the text and then I can finalise this.

**Comment by jarioksa** [comment]
This was more complicated than I expected (but I should have learnt to expect complications). At the first look this looks good, but I'll run more extensive tests tomorrow.

**Comment by gavinsimpson** [comment]
@jarioksa OK; I'll push this to a branch so you can check when you get time.

**Comment by gavinsimpson** [comment]
@jarioksa I have pushed the changes shown above to the [`issue-682`](https://github.com/vegandevs/vegan/tree/issue-682) branch

**Comment by jarioksa** [comment]
Gav,

I’ll have a look at these changes “asap” – but have you heare the Danish word “hygge”? For me this means that the “asap” could be Monday. I could imagine some potential complicataions. For insance, you may have aliased partial terms (Condition). Thet are aliased, but you won’t see none of them – aliiased nor non-aliased. What would happen with overfitted model with matrix arguments, such as having pcnm matix as Conditon or constraints – or how rda(y, x, z) without formula would be handled. These are the things I would look at aong normal tests. But perhaps first on Monday…

Cheers, J

**Comment by jarioksa** [comment]
Gav,

I have played with vegan and tried different cases. All seem to work well. I think this can be merged to master, and I think it could go to cran-2.6-6.

Cheers, Jari

**Comment by jarioksa** [comment]
Gav, I pushed one more commit to issue-682 branch (db5cfc2f0): man/plot.cca.Rd builds automatically an entry in the manual, and this printed a confusing looking message when building vegan (or when generating man page with `R CMD Rd2txt man/plot.cca.Rd`).

If you decide to push these changes to cran-2.6-6, do not cherry-pick this change. man/plot.cca.Rd files have diverged so widely between master and cran-2.6-6 that cherry-picking would fail – and this change is not needed because cran-2.6-6 does not use this trick to build the data base dynamicall

**Comment by jarioksa** [comment]
One unfortunate consequence of being loud is that you may be too boorish and shout all the time. If an ordination command is repeated in a loop, you may be flooded with messages of aliased variables. I already tackled this with `ordistep` in 1ff34bbe68ebe79a8978103d34e6a3f689cf9b87, but there are other **vegan** functions which need attention. Probably several dependent packages have the same annoying feature. Incidentally I saw this in `BiodiversityR::nested.anova.dbrda`, but there are many other in many other packages (and even in **vegan**).

**Comment by gavinsimpson** [comment]
Yeah; we discussed this immediate message and thought it helpful to highlight the problem of aliasing because too many people were not seeing the other messages we print.

For whom should we develop vegan; our users or other package devs? It's pretty easy for the latter to suppress these messages, though it needs those package authors to make changes. But we could also escalate this to a warning as the default behaviour is to gather them and print once execution has ended. This issue deserves a warning I think, to draw attention to the problem.

Personally, I'd just leave this as it is now and let the other developers deal with it. Or we move this to a warning.

**Comment by jarioksa** [comment]
Although I hinted to `warning()`, it is out of question. There is no error or warning state, but just a normal and correct behaviour - we only need to make users aware of this correct behaviour (and yes, I agree, and we have seen that they must be made aware). `warning()` would also alert CRAN checkers, and that is something we do not want to do unnecessarily. I haven't found any way to reduce numbers of messages from one command, but they are printed immediately.

**Comment by jarioksa** [comment]
@gavinsimpson I saw that you already discussed this looping issue in your message of August 16, 2024. I'll have a look at these in **vegan**. I tackled `add1.cca` and `drop1.cca` in `ordistep`, but got to see what would be the best place to handle repeating messages.

**Comment by jarioksa** [comment]
I have now removed all the annoying messages from internal steps that I found. It may be that this makes dependent packages to work more silently, too.

**Comment by jarioksa** [comment]
It seems that people need to have a warning also when conditions completely alias constraints, and constrained component (CCA) is not even returned with the model. The model formula has constraints, but they do not appear in the result which only has `pCCA` and `CA` components. Not sure how to do this, though. See https://stackoverflow.com/questions/79613784/repeated-measures-in-rda-aka-fixed-or-random-effects-why-does-condition-r

**Comment by jarioksa** [comment]
Run time message is easy. Now somewhere by line 200 we return(NULL) if rank==0 in constrained ordination. OK, but we also should give a message. I know two cases: constraints are completely aliased by conditions, or constraints explain nothing. The latter happens, for instance, when a constraint is constant. Actually, the rank==0 return was just added against constant constraints (commit 0a894d0f8).

Permanent NOTE in `print` is a bit trickier. We should know if `CCA` is missing because it was never defined or because it was redundant (NULL). That information is not directly available in the result object (perhaps if `terms` have items that do not have `Condition(` in them, we had a call with constraints but missing constrained component).

**Comment by jarioksa** [comment]
The lost the warnings in 2.5-1 when we moved `ordConstrained`. Earlier we returned something even when the component was NULL. From constrained component (`CCA`) we used to get `list(eig=0, rank=0, qrank=qrank, tot.chi=0, QR=Q, u=<0-col matrix>, v=<0-col matrix>, wa=<0-col matrix>, biplot=matrix(0,0,0), alias=colnames(Y))`. Completely aliased components could be identified by zero rank or zero eigen. Before 2.5-1 we got in `cca-object.tests.R`:
```r
Call: cca(formula = dune ~ Condition(Management) + Management, data =
dune.env)

              Inertia Proportion Rank
Total          2.1153     1.0000     
Conditional    0.6038     0.2855    3
Constrained    0.0000     0.0000    0
Unconstrained  1.5114     0.7145   16
Inertia is scaled Chi-square 
Some constraints were aliased because they were collinear (redundant)
```
In current development (2.7-0) we have:
```r
Call: cca(formula = dune ~ Condition(Management) + Management, data =
dune.env)

              Inertia Proportion Rank
Total          2.1153     1.0000     
Conditional    0.6038     0.2855    3
Unconstrained  1.5114     0.7145   16

Inertia is scaled Chi-square
```
We have lost the zero-line of constrained component and the note on aliased variables. I am not sure we need to have the all-zero line, but at least we need a NOTE in print-out.

**Comment by jarioksa** [comment]
Merge commit 20c99a9230e5096a1eac040074138005d65c0a4f fixes NOTEs with completely aliased `CCA` component. It also adds back the all-zero line for zero-rank constrained component:

```r
Call: cca(formula = dune ~ Condition(Management) + Management, data =
dune.env)

              Inertia Proportion Rank
Total          2.1153     1.0000     
Conditional    0.6038     0.2855    3
Constrained    0.0000     0.0000    0
Unconstrained  1.5114     0.7145   16

Inertia is scaled Chi-square

-- NOTE:
Some constraints or conditions were aliased because they were redundant.
This can happen if terms are constant or linearly dependent (collinear):
‘ManagementHF’, ‘ManagementNM’, ‘ManagementSF’
```

**Comment by jarioksa** [comment]
Remaining issue of messaging on completely aliased components were fixed with merge commits 20c99a9230e5 and 5e7fc999f394.
